### PR TITLE
Allow pipe from input stream to OutgoingMessage to work properly

### PR DESCRIPTION
**Pull request by bainian12345**
## Description

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x] If applicable, the public code is properly documented.
- [x] The code builds without any errors.

### CODE REVIEW

File: hyco-https/lib/_hyco_outgoing.js [hunk 1]
```diff
@@ -659,19 +674,22 @@ OutgoingMessage.prototype._assignSocket = function _assignSocket() {
...
+  // this._forceRendezvousTimer = null;
   if (this._rendezvousChannel) {
     // A rendezvous socket has been assigned
     webSocket = this._rendezvousChannel;
     webSocket._httpMessage = this;
     webSocket.on('close', onServerResponseClose);
   } else {
     // No socket assigned yet, check the message size to decide between the control channel or a new rendezvous
-    if (shouldCreateRendezvous(this.outputSize)) {
+    if (isBufferSizeExceeded(this.outputSize) || this._shouldForceRendezvous) {
```
Comment (dlstucki) [inline hyco-https/lib/_hyco_outgoing.js]: >this._shouldForceRendezvous [](start = 49, length = 27)

This is the commented out field, right? Is that a problem? #Closed

File: hyco-https/lib/_hyco_outgoing.js [hunk 2]
```diff
@@ -784,25 +802,24 @@ OutgoingMessage.prototype._flush = function _flush() {
 };
 
 OutgoingMessage.prototype._flushOutput = function _flushOutput(socket) {
-  var ret;
   var outputLength = this.output.length;
   if (outputLength <= 0)
-    return ret;
```
Comment (dlstucki) [inline hyco-https/lib/_hyco_outgoing.js:790]: >ret [](start = 11, length = 3)

Well that's weird (the previous code). #Closed

File: hyco-https/lib/_hyco_outgoing.js [hunk 3]
```diff
@@ -604,9 +606,22 @@ function write_(msg, chunk, encoding, callback, fromEnd) {
   // signal the user to keep writing.
   if (chunk.length === 0) return true;
 
+
+  // When a input stream pipes to our OutgoingMessage, pipe will call this write function
+  // Returning false will cause the input stream to pause until the 'drain' event is emitted
+  // Therefore we must emit the 'drain' event AFTER this write function returns, which is why this Promise is used.
+  var writeResolve;
+  var writePromise = new Promise(function(resolve, reject) {
```
Comment (dlstucki) [inline hyco-https/lib/_hyco_outgoing.js:614]: >writePromise [](start = 6, length = 12)

Can we talk in person and have you explain to me what this does? #Closed

File: hyco-https/lib/_hyco_outgoing.js (hunk 1 above)
Comment (bainian12345) [reply hyco-https/lib/_hyco_outgoing.js]: Ooops, this should not be there

---
In reply to: [340264312](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340264312) [](ancestors = 340264312)

File: hyco-https/tests/pipe.test.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+var key = process.env.SB_HC_KEY ? process.env.SB_HC_KEY.replace(/^"(.*)"$/, '$1') : null;
+
+expect(ns).toBeDefined();
+expect(path).toBeDefined();
+expect(keyrule).toBeDefined();
+expect(key).toBeDefined();
+
+var smallMessage = "SmallMessage";
+var kb = "";
+for (var i = 1024; i >= 0; i--) {
```
Comment (dlstucki) [inline hyco-https/tests/pipe.test.js]: >(var i = 1024; i >= 0; i-- [](start = 4, length = 26)

Does this do 1025? #Closed

File: hyco-https/tests/pipe.test.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    kb += String.fromCharCode(i % 128);
+}
+// let stream push 1kb at a time
+var over64kbMessage = []; 
+for (var j = 64; j >= 0; j--) {
```
Comment (dlstucki) [inline hyco-https/tests/pipe.test.js:21]: >j = 64; j >= 0 [](start = 9, length = 14)

Does this do 65? Is that intentional? #ByDesign

Comment (bainian12345) [reply hyco-https/tests/pipe.test.js:21]: was intentional to exceed 64kb

---
In reply to: [340271479](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340271479) [](ancestors = 340271479)

File: hyco-https/tests/pipe.test.js (hunk 4 above)
Comment (bainian12345) [reply hyco-https/tests/pipe.test.js]: intentional to exceed 64kb

---
In reply to: [340269407](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340269407) [](ancestors = 340269407)

File: hyco-https/tests/pipe.test.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    over64kbMessage.push(kb);
+}
+
+jest.setTimeout(3000);
```
Comment (dlstucki) [inline hyco-https/tests/pipe.test.js]: >3000 [](start = 16, length = 4)

Does this mean the test must run in under 3 seconds? Would 5 or 10 allow the test to succeed in scenarios where this test is good but the machine is having issues (external to the tests)? #Closed

File: hyco-https/lib/_hyco_outgoing.js (hunk 2 above)
Comment (bainian12345) [reply hyco-https/lib/_hyco_outgoing.js:790]: ret = socket.send() actually returns null all the time, no idea why it was there

---
In reply to: [340264737](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340264737) [](ancestors = 340264737)

File: hyco-https/tests/pipe.test.js (hunk 4 above)
Comment (dlstucki) [reply hyco-https/tests/pipe.test.js]: I guess naming it "kb" though it's kbPlusOne confused me.

---
In reply to: [340272075](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340272075) [](ancestors = 340272075,340269407)

File: hyco-https/tests/pipe.test.js (hunk 6 above)
Comment (bainian12345) [reply hyco-https/tests/pipe.test.js]: That would be good idea

---
In reply to: [340272225](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340272225) [](ancestors = 340272225)

File: hyco-https/tests/pipe.test.js (hunk 5 above)
Comment (dlstucki) [reply hyco-https/tests/pipe.test.js:21]: Ah, I should have noticed the name 'over64kbMessage'.

---
In reply to: [340271956](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340271956) [](ancestors = 340271956,340271479)

File: hyco-https/tests/pipe.test.js (hunk 4 above)
Comment (bainian12345) [reply hyco-https/tests/pipe.test.js]: Good point, will change it

---
In reply to: [340272075](https://github.com/Azure/azure-relay-node/pull/44#discussion_r340272075) [](ancestors = 340272075,340269407)

**Review by dlstucki — APPROVED** [review]
:shipit:

**Pull request merged**